⚡ Faster `decomposeFloat/Double`

zbjornson opened the pull request:
This replaces the (creative) loop-based implementation with a constant-time one. The number 1.234 is about 1100x faster and 1e30 is about 2300x faster.

If you're okay with this change I can do the same thing for double.

<!-- Type of PR: [ ] unchecked / [ ] checked -->

**_Category:_**

- [ ] ✨ Introduce new features
- [ ] 📝 Add or update documentation
- [ ] ✅ Add or update tests
- [ ] 🐛 Fix a bug
- [ ] 🏷️ Add or update types
- [x] ⚡️ Improve performance
- [ ] _Other(s):_ ...
  <!-- Don't forget to add the gitmoji icon in the name of the PR -->
  <!-- See: https://gitmoji.dev/                                  -->

<!-- Fixing bugs, adding features... may impact existing ones           -->
<!-- in order to track potential issues that could be related to your PR -->
<!-- please check the impacts and describe more precisely what to expect -->

**_Potential impacts:_**

<!-- Generated values: Can your change impact any of the existing generators in terms of generated values, if so which ones? when? -->
<!-- Shrink values:    Can your change impact any of the existing generators in terms of shrink values, if so which ones? when? -->
<!-- Performance:      Can it require some typings changes on user side? Please give more details -->
<!-- Typings:          Is there a potential performance impact? In which cases? -->

- [ ] Generated values
- [ ] Shrink values
- [x] Performance - faster
- [ ] Typings
- [ ] _Other(s):_ ...

dubzzz commented:
Looks awesome. I'll spend some time checking the diff but I think we can go with a similar optim for doubles

zbjornson commented:
Thanks! I'm happy to work on the double variant also.

It looks like I broke something. How do you get unit tests to run locally? If I run `yarn workspace fast-check test`, I get a ton of TS errors.

dubzzz commented on packages/fast-check/src/arbitrary/_internals/helpers/FloatHelpers.ts:
```diff
@@ -17,12 +16,18 @@ const INDEX_POSITIVE_INFINITY = 2139095040; // floatToIndex(MAX_VALUE_32) + 1;
 /** @internal */
 const INDEX_NEGATIVE_INFINITY = -2139095041; // floatToIndex(-MAX_VALUE_32) - 1
 
+const buffer = new ArrayBuffer(4);
+const view = new DataView(buffer);
```
**Perfect! It even works with node 8** which is the currently minimal supported version for fast-check (will be dropped soon, but so far v3 is supposed to support it).

dubzzz commented:
You may try:

```sh
yarn # install all deps
yarn build:all # build all packages

# usually I changedir into /packages/fast-check
# but workspace commands also work
yarn test
yarn e2e # runs against prod bundle so need `yarn build` to be called before if you change things
```

You'll also need to declare the version bump (minor for fast-check and none for others):

```sh
yarn version:bump
```

_I'd need to update my contributing guide with this setup_

zbjornson commented:
Thanks for the guidance there. Fixed the failing test, switched from `DataView` to `TypedArray` for slightly better perf, ran `version:bump`, and added another commit for `decomposeDouble`.

dubzzz commented:
It seems that format is invalid. Can you please run `yarn format` on your PR?

dubzzz commented:
Thank you so much for this awesome contribution 🥰

zbjornson commented:
My pleasure, thanks for the fast review.
